drivers: spi: support for Nuvoton numaker SPI interface

cyliangtw opened the pull request:
This PR is to add Nuvoton numaker series SPI interface.
It's verified on board: NuMaker-PFM-M467.

fabiobaltieri commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,358 @@
+static struct spi_driver_api spi_numaker_driver_api = {.transceive = spi_numaker_transceive,
+						       .release = spi_numaker_release};
+
+int spi_numaker_init(const struct device *dev)
```
static

cyliangtw replied:
Done

fabiobaltieri commented on tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:
```diff
@@ -0,0 +1,34 @@
+	status = "okay";
+	pinctrl-0 = <&spi2_default>;
+	pinctrl-names = "default";
+/*    cs-gpios = <&gpioa 11 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>; */
```
Drop it?

fabiobaltieri replied:
If it's for the loopback test, why not just enabling it? There's plenty of tests like this that just picks a pin that makes sense so it's ready to use.

cyliangtw replied:
Agree with your opinion to drop this comment.

fabiobaltieri commented on tests/drivers/spi/spi_loopback/boards/numaker_pfm_m467.overlay:
```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+&pinctrl {
+/* CS/CLK/MISO/MOSI --> EVB's NU5: SS/CLK/MISO/MOSI --> PA11/PA10/PA9/PA8 */
```
Do you really need the comment? The pin names have that information already, this is one more thing to get stale if you end up copypasting it on new boards. I'd suggest dropping it.

cyliangtw replied:
Could I only keep `/* EVB's NU5: SS/CLK/MISO/MOSI */` ?
else user can't easy find the pad "NU5" to flywire MISO/MOSI for SPI loopback test.

fabiobaltieri replied:
Oh yeah that makes sense.

cyliangtw commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,358 @@
+	}
+
+	/* full/half duplex */
+	if (((config->operation) & BIT(11))) {
```
Replace by meaningful name SPI_HALF_DUPLEX & SPI_CS_ACTIVE_HIGH.

cyliangtw commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,358 @@
+			  (mode == 0)   ? QSPI_MODE_0
+			  : (mode == 1) ? QSPI_MODE_1
+			  : (mode == 2) ? QSPI_MODE_2
+					: QSPI_MODE_3,
```
Done

cyliangtw commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,358 @@
+	const struct spi_numaker_config *dev_cfg = dev->config;
+	struct spi_context *ctx = &data->ctx;
+	uint32_t tx_frame = 0U, rx_frame = 0U;
+	uint8_t word_size = 0, spi_dfs = 0;
```
yes, it's not necessary.

cyliangtw commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,358 @@
+		break;
+	default:
+		spi_dfs = 0;
+		ret = -1;
```
Refined and finished the update at commit  532e5e4aa72c

fabiobaltieri commented on drivers/spi/spi_numaker.c:
```diff
@@ -0,0 +1,351 @@
+ * CPOL/CPHA = 1/0 --> SPI_MODE_2
+ * CPOL/CPHA = 1/1 --> SPI_MODE_3
+ */
+const uint32_t smode_tbl[4] = {
```
`static` (below as well)

cyliangtw replied:
Update at commit c7392e8f739

cyliangtw committed:
drivers: spi: support for Nuvoton numaker series

Add Nuvoton numaker series spi controller, including
full and half duplex support.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw committed:
tests: drivers: spi: spi_loopback: support numaker_pfm_m467

Add support for Nuvoton numaker board numaker_pfm_m467.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw commented:
Hi @tbursztyka,
please take a look.

cyliangtw commented:
Hi @tbursztyka , @ssekar15 ,
If you are available, please look into this PR.

ssekar15 reviewed (approved):
thanks, looks fine for me
